### ISSUE TITLE
Sound null safety disabled in flutter web release build

### ISSUE THREAD

**Issue by amoffatgmi** [flutter/flutter#113814]
It appears that sound null safety is disabled during a web release build. This was unexpected and caused a bug in my application when `jsonDecode` was used (which produces a `dynamic`) and a key didn't exist in the returned `dynamic`. A `null` value was assigned to a non-nullable type, which caused a downstream bug. I was unable to produce this failure during debugging, which is when I realized that sound null safety appears to be switched off in release but not debug.

If there is a performance reason for this, I would request that a commandline option allow sound null safety to remain enabled.

## Steps to Reproduce

1. Create any sample flutter app.
2. In a widget's build method, cast a `Map` to a `dynamic` and access a non-existent key.
3. Store the value of the non-existent key in a non-null String variable.
4. Display the value of the non-null String variable to the screen.
5. Run the app in web debug mode, observe behavior.
6. Run the app in web release mode, observe behavior.

**Expected results:** 

Debug build:
![image](https://user-images.githubusercontent.com/102402443/197119623-daceaf77-488b-47ac-9329-429fcce790d5.png)

Release build:
![image](https://user-images.githubusercontent.com/102402443/197119623-daceaf77-488b-47ac-9329-429fcce790d5.png)

**Actual results:** 

Debug build:
![image](https://user-images.githubusercontent.com/102402443/197119623-daceaf77-488b-47ac-9329-429fcce790d5.png)

Release build:
![image](https://user-images.githubusercontent.com/102402443/197119767-38b3572d-b083-4ddd-9cc5-2c0d3eb77bef.png)

<details>
<summary>Code sample</summary>

```dart
  @override
  Widget build(BuildContext context) {
   String thing = ({} as dynamic)["thing"];
   return Text("thing: $thing");
  }
```

</details>

<details>
  <summary>Logs</summary>

```
Launching lib\main.dart on Chrome in debug mode...
This app is linked to the debug service: ws://127.0.0.1:65461/bPBQlViR65o=/ws
Debug service listening on ws://127.0.0.1:65461/bPBQlViR65o=/ws
 Running with sound null safety
Connecting to VM Service at ws://127.0.0.1:65461/bPBQlViR65o=/ws
Flutter Web Bootstrap: Programmatic
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following TypeErrorImpl was thrown building MyHomePage(dirty, state: _MyHomePageState#a05ad):
Expected a value of type 'String', but got one of type 'Null'
The relevant error-causing widget was:
MyHomePage
When the exception was thrown, this was the stack:
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 266:49      throw_
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 99:3        castError
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 452:10  cast
C:/b/s/w/ir/cache/builder/src/out/host_debug/dart-sdk/lib/_internal/js_dev_runtime/patch/core_patch.dart 705:17                as
packages/null_sound_safety_bug/main.dart 68:12                                                                                 build
packages/flutter/src/widgets/framework.dart 4992:27                                                                            build
...
```

```
[√] Flutter (Channel stable, 3.3.4, on Microsoft Windows [Version 10.0.22621.674], locale en-US)
    • Flutter version 3.3.4 on channel stable at C:\Users\**********\Documents\Work\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision eb6d86ee27 (2 weeks ago), 2022-10-04 22:31:45 -0700
    • Engine revision c08d7d5efc
    • Dart version 2.18.2
    • DevTools version 2.15.0

[√] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at C:\Users\**********\AppData\Local\Android\Sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.3.4)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.3.32901.215
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2021.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+7-b1504.28-7817840)

[√] VS Code (version 1.72.2)
    • VS Code at C:\Users\**********\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.50.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22621.674]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 106.0.5249.119
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 105.0.1343.53

[√] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by mraleph** [comment]
In release builds Flutter tooling invokes `dart2js` with `-O4` optimization level which discards all implicit downcasts. 

```
 -O3
       Enables optimizations that respect the language semantics only on
       programs that do not ever throw any subtype of `Error`.  These
       optimizations improve the generated code, but they may cause programs to
       behave unexpectedly if this assumption is not met.  To use this
       option, we recommend that you properly test your application first
       without it, and ensure that no subtype of `Error` (such as `TypeError`)
       is ever thrown.

       Equivalent to calling dart2js with these extra flags:
         -O2
         --omit-implicit-checks

 -O4
       Enables more aggressive optimizations than -O3, but with the same
       assumptions. These optimizations are on a separate group because they
       are more susceptible to variations in input data. To use this option we
       recommend to pay special attention to test edge cases in user input.

       Equivalent to calling dart2js with these extra flags:
         -O3
         --trust-primitives

  --omit-implicit-checks
    Omit implicit runtime checks, such as parameter checks and implicit
    downcasts. These checks are included by default in Dart 2.0. By
    using this flag the checks are removed, however the compiler will assume
    that all such checks were valid and may use this information for
    optimizations. Use this option only if you have enough testing to ensure
    that your program works with the checks.
```

You could use `--dart-define=Dart2jsOptimization=O2` to switch to a lower optimization level though performance would suffer.

A better solution would be to rewrite your code to use explicit casts rather than implicit. `dart2js` will not remove explicit casts, which means the code below will actually throw as you expect: 

```dart
  final thing = ({} as dynamic)["thing"] as String;
```

To find all the places where you rely on implicit casts from `dynamic` you can enable `strict-casts` in the analyzer as described [here](https://dart.dev/guides/language/analysis-options#enabling-additional-type-checks), by adding the following to your `analysis_options.yaml`:

```
analyzer:
  language:
    strict-casts: true
```

**Comment by amoffatgmi** [comment]
I like the explicit casting solution a lot. Thanks for explaining @mraleph and also for the workarounds.
